lang/LangPrimSource/SC_AudioDevicePrim.cpp

```diff
@@ -125,6 +131,56 @@ int listDevices(struct VMGlobals* g, int type) {
     return 1;
 }
 
+#elif defined(SC_AUDIO_API_PORTAUDIO)
+int listDevices(struct VMGlobals* g, int type) {
+    int numDevices;
+    PaError err = Pa_Initialize();
+    if (err != paNoError)
+        return 0;
+
+    numDevices = Pa_GetDeviceCount();
+    if (numDevices < 0)
+        return 0;
+
+    std::vector<const PaDeviceInfo*> deviceInfos;
+    for (PaDeviceIndex i = 0; i < numDevices; i++) {
+        auto* pdi = Pa_GetDeviceInfo(i);
+        if (type < BOTH) {
+            if (type < IN) {
+                if (pdi->maxOutputChannels > 0)
+                    deviceInfos.push_back(pdi);
+            } else {
+                if (pdi->maxInputChannels > 0)
+                    deviceInfos.push_back(pdi);
+            }
+        } else
+            deviceInfos.push_back(pdi);
+    }
+
+    int numSelectedDevices = deviceInfos.size();
+
+    PyrSlot* a = g->sp - 2;
+    PyrObject* devArray = newPyrArray(g->gc, numSelectedDevices * sizeof(PyrObject), 0, true);
+    SetObject(a, devArray); // this is okay here as we don't use the receiver below
```

Comment:
having a local variable named `a` is not really helping here -- you can just write `SetObject(g->sp - 2, devArray)`